### PR TITLE
COOPR:629 Update template APIs to support extending templates and including partials

### PR DESCRIPTION
**Pull request by rpedanov**
@jwang47 Branch ready for review.

### CODE REVIEW

**Comment by rpedanov** [comment]
@jwang47 I'm not sure why build fails, it looks like some npm problems. Could you check it please?

File: coopr-server/src/main/java/co/cask/coopr/spec/template/Administration.java [hunk 1]
```diff
@@ -20,10 +20,10 @@
 /**
  * Defines settings for cluster administration like lease duration, etc.
  */
-public class Administration {
+public final class Administration {
   public static final Administration EMPTY_ADMINISTRATION = new Administration(LeaseDuration.FOREVER_LEASE_DURATION);
 
-  private final LeaseDuration leaseDuration;
+  LeaseDuration leaseDuration;
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/Administration.java]: Why remove `private final`?

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterDefaults.java [hunk 2]
```diff
@@ -16,32 +16,28 @@
...
+  private String dnsSuffix;
+  private JsonObject config;
 
   private ClusterDefaults(Set<String> services, String provider, String hardwaretype,
                           String imagetype, String dnsSuffix, JsonObject config) {
-    Preconditions.checkArgument(services != null && !services.isEmpty(), "default services must be specified");
-    //TODO: move it to the final phase
-//    Preconditions.checkArgument(provider != null, "default provider must be specified");
-    Preconditions.checkArgument(dnsSuffix == null || StringUtils.isValidDNSSuffix(dnsSuffix),
-                                dnsSuffix + " is an invalid DNS suffix.");
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterDefaults.java:44]: Why remove these checks?

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterDefaults.java [hunk 3]
```diff
@@ -129,12 +157,12 @@ public static Builder builder() {
     private JsonObject config = new JsonObject();
 
     public Builder setServices(Set<String> services) {
-      this.services = ImmutableSet.copyOf(services);
+      this.services = services == null ? Sets.<String>newHashSet() : services;
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterDefaults.java]: `services == null ? ImmutableSet.of() : ImmutableSet.copyOf()`

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterDefaults.java [hunk 4]
```diff
@@ -129,12 +157,12 @@ public static Builder builder() {
...
       return this;
     }
 
     public Builder setServices(String... services) {
-      this.services = ImmutableSet.copyOf(services);
+      this.services = new HashSet<String>(Arrays.asList(services));
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterDefaults.java]: Why change this to be mutable set?

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java [hunk 5]
```diff
@@ -97,6 +110,122 @@ public ClusterTemplate build() {
...
+      private final Set<String> immutables = Sets.newHashSet();
+
+      public Merger setInitialTemplate(ClusterTemplate initialTemplate) {
+        parent = initialTemplate.parent;
+        includes = initialTemplate.includes;
+
+        return this;
+      }
+
+      public Merger merge(Collection<AbstractTemplate> templates) throws TemplateImmutabilityException {
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java]: Should be `List<AbstractTemplate>` since order is important.

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java [hunk 6]
```diff
@@ -97,6 +110,122 @@ public ClusterTemplate build() {
...
+        for (AbstractTemplate template : templates) {
+          if (template instanceof ClusterTemplate) {
+            copyFullProps(template);
+          } else {
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java]: Should be `else if (template Instanceof PartialTemplate)`, and throw in the `else` - if we have a new child of `AbstractTemplate`, then we won't get unexpected behavior.

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java [hunk 7]
```diff
@@ -97,6 +110,122 @@ public ClusterTemplate build() {
...
+        return merge(Arrays.asList(template));
+      }
+
+      private void validateFields() {
+        Preconditions.checkArgument(clusterDefaults.getServices() != null && !clusterDefaults.getServices().isEmpty(),
+                                    "default services must be specified");
+        Preconditions.checkArgument(clusterDefaults.getProvider() != null, "default provider must be specified");
+        Preconditions.checkArgument(clusterDefaults.getDnsSuffix() == null ||
+                                      StringUtils.isValidDNSSuffix(clusterDefaults.getDnsSuffix()),
+                                    clusterDefaults.getDnsSuffix() + " is an invalid DNS suffix.");
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java]: Ok, so the checks are moved here.

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java [hunk 8]
```diff
@@ -73,10 +83,13 @@ public Parent getParent() {
...
    */
   public static class Builder extends AbstractTemplate.Builder<ClusterTemplate, Builder> {
-
     private Parent parent;
     private Set<Include> includes = ImmutableSet.of();
 
+    public Merger merger() {
+      return new Merger();
+    }
+
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java]: I don't think `Merger` should come from a builder method. What if we want to use multiple `Merger` implementations? I think we should instead have `Merger` be injected into whatever classes want to merge templates.

File: coopr-server/src/main/java/co/cask/coopr/spec/template/Compatibilities.java [hunk 9]
```diff
@@ -133,37 +135,37 @@ public static Builder builder() {
    * Builder for creating Compatibilities.
    */
   public static class Builder {
-    private Set<String> hardwaretypes = ImmutableSet.of();
-    private Set<String> imagetypes = ImmutableSet.of();
-    private Set<String> services = ImmutableSet.of();
+    private Set<String> hardwaretypes = Sets.newHashSet();
+    private Set<String> imagetypes = Sets.newHashSet();
+    private Set<String> services = Sets.newHashSet();
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/Compatibilities.java]: Why change these fields to be mutable sets?

File: coopr-server/src/main/java/co/cask/coopr/spec/template/Constraints.java [hunk 10]
```diff
@@ -28,12 +28,12 @@
...
   private final Map<String, ServiceConstraint> serviceConstraints;
   private final LayoutConstraint layoutConstraint;
-  private final SizeConstraint sizeConstraint;
+  SizeConstraint sizeConstraint;
 
   public Constraints(Map<String, ServiceConstraint> serviceConstraints, LayoutConstraint layoutConstraint,
                      SizeConstraint sizeConstraint) {
     this.serviceConstraints = serviceConstraints == null ?
-      ImmutableMap.<String, ServiceConstraint>of() : serviceConstraints;
+      Maps.<String, ServiceConstraint>newHashMap() : serviceConstraints;
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/Constraints.java]: Why change this?

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java [hunk 11]
```diff
@@ -97,6 +110,122 @@ public ClusterTemplate build() {
...
+        //merge base entity properties
+        if (isNotBlank(from.getName())) name = from.getName();
+        if (isNotBlank(from.getLabel())) label = from.getLabel();
+        if (isNotBlank(from.getIcon())) icon = from.getIcon();
+        if (isNotBlank(from.getDescription())) description = from.getDescription();
+      }
+
+      @SuppressWarnings("unchecked")
+      private void mergeSet(Set dest, Set from) {
+        if (dest != null && from != null) {
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java]: We made a lot of things mutable sets due to how we're merging the template here, right?

Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java]: Since we made a lot of things mutable sets/maps, let's move the immutable copy to the `build()` method of the builder, so that a built object doesn't use mutable sets/maps.

File: coopr-server/src/test/java/co/cask/coopr/spec/template/PartialTemplateTest.java [hunk 12]
```diff
@@ -2,23 +2,30 @@
...
+import org.junit.runners.MethodSorters;
 
 import java.io.InputStream;
+import java.util.Map;
+import java.util.Set;
 
 /**
  *
  */
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
```
Comment (jwang47) [inline coopr-server/src/test/java/co/cask/coopr/spec/template/PartialTemplateTest.java:29]: Why is this needed?

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java [hunk 13]
```diff
@@ -97,6 +110,122 @@ public ClusterTemplate build() {
...
+        mergeSet(compatibilities.getHardwaretypes(), from.getCompatibilities().getHardwaretypes());
+        mergeSet(compatibilities.getImagetypes(), from.getCompatibilities().getImagetypes());
+        mergeSet(compatibilities.getServices(), from.getCompatibilities().getServices());
+
+        //merge admin lease duration
+        if (!from.getAdministration().getLeaseDuration().equals(LeaseDuration.FOREVER_LEASE_DURATION)) {
+          administration.leaseDuration = from.getAdministration().getLeaseDuration();
+        }
+
+        //merge base entity properties
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java]: We don't want to merge these. Instead, we'll use whatever is immediately provided. E.g. with user JSON, we would simply use what is provided in that JSON; if it's from a specified template, then it's from that template.

**Comment by jwang47** [comment]
Please resolve merge conflicts

File: coopr-server/src/main/java/co/cask/coopr/spec/template/Administration.java (hunk 1 above)
Comment (rpedanov) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/Administration.java]: For merging purposes. We have to replace leaseDuration at the final merging process.

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterDefaults.java (hunk 2 above)
Comment (rpedanov) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterDefaults.java:44]: Since we have partial templates this fields can be empty. So I moved this checks into Merger.validateFields() that it can be validated when the template will be merged.

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterDefaults.java (hunk 3 above)
Comment (rpedanov) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterDefaults.java]: Since we have to merge several templates we have to merge all inner collections of this templates. For example if we have two templates with "defaults" for merge then we have to merge "services" set from second defaults into the first one. If set will be immutable then this operation will be no successfully.

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterDefaults.java (hunk 4 above)
Comment (rpedanov) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterDefaults.java]: See comment above

File: coopr-server/src/test/java/co/cask/coopr/spec/template/PartialTemplateTest.java (hunk 12 above)
Comment (rpedanov) [inline coopr-server/src/test/java/co/cask/coopr/spec/template/PartialTemplateTest.java:29]: For test ordering. I want that templates will be parsed and persisted at the begin so we can get it in the next tests, and deleted at the final test method.

File: coopr-server/src/main/java/co/cask/coopr/spec/template/Constraints.java (hunk 10 above)
Comment (rpedanov) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/Constraints.java]: For merging purposes. See comment above.

File: coopr-server/src/main/java/co/cask/coopr/spec/template/Compatibilities.java (hunk 9 above)
Comment (rpedanov) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/Compatibilities.java]: For merging purposes. See comment above.

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java (hunk 13 above)
Comment (rpedanov) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java]: Now this fields will be set with the values of last template that be provided in merge set. I mean that if we have something like that:
Child -> Parent -> Includes
Then values of this fields will be replaced with child values because we merge child at final. This allow to avoid some null value from the child. For example if parent has Label field value and child has not, then Label will be taken from parent. Or we should take this fields only from the last template(specified template) and set null values if values does not provided?

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java (hunk 5 above)
Comment (rpedanov) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java]: LinkedHashSet always passed here. I believe that it saves insert order as well.

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java (hunk 13 above)
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java]: Let's take these fields only from the last template.

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java (hunk 11 above)
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java]: @rpedanov Please look at this comment.

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java (hunk 8 above)
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java]: @rpedanov Please look at this comment too.

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java (hunk 11 above)
Comment (rpedanov) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java]: Hey @jwang47 , as I understand I have to remove all Mutable collections and its setters from builders? But create setters for mutable collections inside template entities (ClusterDefaults, Compatibilities)?  So if we try to create some entity we have to construct it from builder first and after that init not mutable fields through setter of this entity?

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java (hunk 8 above)
Comment (rpedanov) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java]: @jwang47 Yes, I already did it. I'm waiting when I done with moving mutable collections. Please, see my question below.

File: coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java (hunk 11 above)
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java]: Right, let's keep the collections of the template entities (ClusterDefaults, Compatibilities) immutable as they were before. Just don't want to allow the possibility of any unexpected bugs.

**Comment by rpedanov** [comment]
Immutable collections came back to template entities, merger separate service now. @jwang47 please check it.

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 14]
```diff
@@ -732,4 +741,85 @@ private Cluster getCluster(String clusterId, Account account) throws IOException
     }
     return cluster;
   }
+
+  /**
+   * Build cluster template from provided includes and parents.
+   *
+   * @param clusterTemplate Cluster template which can contains includes and parents.
+   * @param account Account of the user that is trying to resume a cluster job.
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: Account of the user that is trying to resolve a template.

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 15]
```diff
@@ -732,4 +741,85 @@ private Cluster getCluster(String clusterId, Account account) throws IOException
...
+    throws TemplateNotFoundException, TemplateImmutabilityException, IOException {
+    EntityStoreView entityStore = entityStoreService.getView(account);
+    return resolveTemplate(entityStore, clusterTemplate);
+  }
+
+  /**
+   * Build cluster template from provided includes and parents.
+   *
+   * @param templateName Cluster template name.
+   * @param account Account of the user that is trying to resume a cluster job.
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: Account of the user that is trying to resolve a template.

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 16]
```diff
@@ -732,4 +741,85 @@ private Cluster getCluster(String clusterId, Account account) throws IOException
...
     return cluster;
   }
+
+  /**
+   * Build cluster template from provided includes and parents.
+   *
+   * @param clusterTemplate Cluster template which can contains includes and parents.
+   * @param account Account of the user that is trying to resume a cluster job.
+   * @return Cluster Template with merged body from includes and parents.
+   * @throws IOException if there was some error writing to stores.
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java:751]: We don't write to stores here, right? Just read?

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 17]
```diff
@@ -732,4 +741,85 @@ private Cluster getCluster(String clusterId, Account account) throws IOException
...
+   * @throws TemplateNotFoundException if template can't be found in store.
```
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: The template is provided, so why do we need to throw TemplateNotFoundException?

**Comment by jwang47** [comment]
Let's also add test cases for user providing an entire template rather than using a stored template. Can test the following:
- Provided template without error
- Provided template that tries to override immutable partial
- Provided template that tries to override immutable parent template
- Provided template that is badly formatted (i.e. not proper JSON)
- etc

(also make sure we have the same set of test cases that use stored templates)

Otherwise, the code changes mostly looks good.

**Comment by rpedanov** [comment]
@jwang47 please take a look

**Comment by jwang47** [comment]
Just one minor comment.

**Comment by rpedanov** [comment]
@jwang47 new line added
